include/albatross/src/tune/greedy_tuner.hpp

```diff
@@ -0,0 +1,208 @@
+    const auto ratios = query_ratios(2 * n);
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value + high_range * ratios[i]);
+    }
+  } else if (value == high) {
+    // only search lower since we're at the upper bound
+    const auto ratios = query_ratios(2 * n);
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value - low_range * ratios[ratios.size() - 1 - i]);
+    }
+  } else {
+    // search in both directions
+    const auto ratios = query_ratios(n);
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value - low_range * ratios[ratios.size() - 1 - i]);
+    }
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value + high_range * ratios[i]);
+    }
+  }
+
+  return queries;
+}
+
+inline ParameterStore set_tunable_param(const ParameterStore &params,
+                                        std::size_t i, double val) {
+  auto perturbed = get_tunable_parameters(params);
+  perturbed.values[i] = val;
+  return set_tunable_params_values(params, perturbed.values, true);
+};
+} // namespace details
+
+template <typename Function>
+inline ParameterStore
+greedy_tune(Function evaluate_function, const ParameterStore &params,
+            std::size_t n_threads = 8, std::size_t n_iterations = 10,
+            bool use_async = true, std::ostream *os = &std::cout) {
+
+  static_assert(
+      has_call_operator<Function, ParameterStore>::value,
+      "evaluate_function must have a single ParameterStore argument.");
+
+  assert(n_threads % 2 == 0); // n_threads must be even;
+  const std::size_t n_queries = n_threads / 2;
+
+  albatross::TunableParameters tunable = get_tunable_parameters(params);
+
+  if (os) {
+    (*os) << "Will be tuning the following:" << std::endl;
+    for (std::size_t i = 0; i < tunable.names.size(); ++i) {
+      (*os) << "    " << tunable.names[i] << "    [ ";
+      const auto queries =
+          details::get_queries(tunable.values[i], tunable.lower_bounds[i],
+                               tunable.upper_bounds[i], n_queries);
+      for (const auto &v : queries) {
+        (*os)
+            << details::set_tunable_param(params, i, v)[tunable.names[i]].value
+            << ", ";
+      }
+      (*os) << " ]" << std::endl;
+    }
+    (*os) << "Initial params:" << std::endl;
+    (*os) << albatross::pretty_params(params) << std::endl;
+  }
+
+  double best_value = HUGE_VAL;
+  ParameterStore best_params(params);
+
+  for (std::size_t iter = 0; iter < n_iterations; ++iter) {
+    for (std::size_t i = 0; i < tunable.names.size(); ++i) {
+
+      tunable = get_tunable_parameters(best_params);
+      auto values =
+          details::get_queries(tunable.values[i], tunable.lower_bounds[i],
+                               tunable.upper_bounds[i], n_queries);
+      if (iter == 0 && i == 0) {
+        // on the very first iteration we need to include the initial params
```

Comment:
What makes iteration 1 unique here? As `best_params` and `tunable` change in value iteration to iterations, wouldn't we always want to keep the surviving best value. As is, would this allow an initial hardcoded value to survive for one round, but (assuming the value won iteration 1), only to be excluded from the search if `get_queries` doesn't include the value?

Reply:
The algorithm starts without knowledge of the current function value, so on the first iteration we evaluate at all the proposed values AND at the initial one.  The following logic which picks the best params then keeps track of the current best which may mean it just selects the initial parameters (if all the proposed perturbations were worse).  By iteration two we're guaranteed to have a candidate for `best_params` so nothing get's excluded.

An alternative approach would be to just evaluate the function with the initial params at the beginning of the algorithm ... but if the function takes an hour to evaluate that means optimization won't start for at least an hour, this is meant to get the processes going sooner.